integration/kotlinx-coroutines-firebase-android/kotlinx-coroutines-firebase-android/src/main/java/kotlinx/coroutines/experimental/firebase/android/Task.kt

```diff
@@ -0,0 +1,55 @@
+package kotlinx.coroutines.experimental.firebase.android
+
+import com.google.android.gms.tasks.Task
+import kotlinx.coroutines.experimental.suspendCancellableCoroutine
+import kotlin.coroutines.experimental.suspendCoroutine
+
+/**
+ * Coroutine support to Firebase Task interface
+ *
+ * This method is not intended to be used in your code. You should probably use [await] sinsce it's
+ * more idiomatic.
+ *
+ * The implementation is pretty simple. It justs use a [suspendCoroutine] to encapsulate the
+ * Firebase [com.google.android.gms.tasks.OnCompleteListener] interface.
+ *
+ */
+private suspend fun <T> awaitTask(task: Task<T>): T = suspendCancellableCoroutine { continuation ->
+    task
+            .addOnSuccessListener(continuation::resume)
+            .addOnFailureListener(continuation::resumeWithException)
+            .addOnCanceledListener { continuation.cancel() }
```

Comment:
This code will produce four unnecessary classes (see decompiled code).
It's better to introduce a separate class such as `Future.kt#ContinuationConsumer`

Reply:
Changed in https://github.com/Kotlin/kotlinx.coroutines/pull/426/commits/2d272a8624ba72a7fd3ab0855604f5103e5c702f.